airflow/providers/cncf/kubernetes/operators/custom_object_launcher.py

```diff
@@ -0,0 +1,333 @@
+from copy import deepcopy
+from datetime import datetime as dt
+from typing import Optional
+
+import tenacity
+from kubernetes import client
+from kubernetes.client import models as k8s
+from kubernetes.client.rest import ApiException
+
+from airflow.exceptions import AirflowException
+from airflow.providers.cncf.kubernetes.utils.pod_manager import PodManager
+from airflow.utils.log.logging_mixin import LoggingMixin
+
+from airflow.providers.cncf.kubernetes.resource_convert.secret import convert_secret, convert_image_pull_secrets
+from airflow.providers.cncf.kubernetes.resource_convert.configmap import convert_configmap, convert_configmap_to_volume
+from airflow.providers.cncf.kubernetes.resource_convert.env_variable import convert_env_vars
+
+if sys.version_info >= (3, 8):
+    from functools import cached_property
+else:
+    from cached_property import cached_property
+
+
+def should_retry_start_spark_job(exception: BaseException) -> bool:
+    """Check if an Exception indicates a transient error and warrants retrying"""
+    if isinstance(exception, ApiException):
+        return exception.status == 409
+    return False
+
+
+class SparkJobSpec():
+    def __init__(self, **entries):
+        self.__dict__.update(entries)
+        self.validate()
+        self.update_resources()
+
+    def validate(self):
+        if self.spec['dynamicAllocation']['enabled']:
+            if not all(
+                [self.spec['dynamicAllocation']['initialExecutors'], self.spec['dynamicAllocation']['minExecutors'], self.spec['dynamicAllocation']['maxExecutors']]
+            ):
+                raise AirflowException("Make sure initial/min/max value for dynamic allocation is passed")
+
+    def update_resources(self):
+        spark_resources = SparkResources(self.spec['driver'].pop('container_resources'), self.spec['executor'].pop('container_resources'))
+        self.spec['driver'].update(spark_resources.resources['driver'])
+        self.spec['executor'].update(spark_resources.resources['executor'])
+
+
+class KubernetesSpec():
+    def __init__(self, **entries):
+        self.__dict__.update(entries)
+        self.set_attribute()
+
+    def set_attribute(self):
+        self.env_vars = convert_env_vars(self.env_vars) if self.env_vars else []
+        self.image_pull_secrets = convert_image_pull_secrets(image_pull_secrets) if self.image_pull_secrets else []
+        if self.config_map_mounts:
+            vols, vols_mounts = convert_configmap_to_volume(self.config_map_mounts)
+            self.volumes.extend(vols)
+            self.volume_mounts.extend(vols_mounts)
+        if self.from_env_config_map:
+            self.env_from.extend([convert_configmap(c_name) for c_name in self.from_env_config_map])
+        if self.from_env_secret:
+            self.env_from.extend([convert_secret(c) for c in self.from_env_secret])
+
+
+class SparkResources:
+    """spark resources"""
+
+    def __init__(
+        self,
+        driver=None,
+        executor=None
```

Comment:
We need typing for this and probably more description of what driver/executor should contain in the Doc string of this class

Reply:
done in [228ce7e](https://github.com/apache/airflow/pull/22253/commits/228ce7ee602132e51544032d23fc47f5720177a1)